pkg/controller/noderef/noderef_controller.go

```diff
@@ -0,0 +1,227 @@
+	cluster, err := r.getCluster(ctx, machine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Warningf("Cannot find a Cluster for Machine %q in namespace %q, won't reconcile", machine.Name, machine.Namespace)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't have a valid ProviderID, retrying later", machine.Name, machine.Namespace)
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	result, err := r.reconcile(ctx, cluster, machine)
+	if err != nil {
+		klog.Errorf("Failed to assign NodeRef to Machine %q: %v", request.NamespacedName, err)
+		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return result, err
+	}
+
+	klog.Infof("Set Machine's (%q in namespace %q) NodeRef to %q", machine.Name, machine.Namespace, machine.Status.NodeRef.Name)
+	r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+	return result, nil
+}
+
+func (r *ReconcileNodeRef) reconcile(ctx context.Context, cluster *v1alpha1.Cluster, machine *v1alpha1.Machine) (reconcile.Result, error) {
+	providerID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterClient, err := remote.NewClusterClient(r.Client, cluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	corev1Client, err := clusterClient.CoreV1()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Get the Node reference.
+	nodeRef, err := r.getNodeReference(corev1Client, providerID)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			klog.Warningf("Cannot find a matching Node for Machine %q in namespace %q, retrying later", machine.Name, machine.Namespace)
+			return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Update Machine.
+	machine.Status.NodeRef = nodeRef
+	if err := r.Client.Status().Update(ctx, machine); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNodeRef) getCluster(ctx context.Context, machine *v1alpha1.Machine) (*v1alpha1.Cluster, error) {
+	cluster := &v1alpha1.Cluster{}
+	key := client.ObjectKey{
+		Namespace: machine.Namespace,
+		Name:      machine.Labels[v1alpha1.MachineClusterLabelName],
+	}
+
+	if err := r.Client.Get(ctx, key, cluster); err != nil {
+		return nil, err
+	}
+
+	return cluster, nil
+}
+
+func (r *ReconcileNodeRef) getNodeReference(client corev1.NodesGetter, providerID *noderefutil.ProviderID) (*apicorev1.ObjectReference, error) {
+	listOpt := metav1.ListOptions{}
+
+	for {
+		nodeList, err := client.Nodes().List(listOpt)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, node := range nodeList.Items {
```

Comment:
In a follow up we could use an indexer here and filter the request by node.Spec.ProviderID

Reply:
@enxebre if you'd like to see that happen, please file an issue. Otherwise, I fear this comment will got lost. Thanks!

Reply:
sure, I'll also put a PR